### PR TITLE
[prometheus-thanos] Make the cache for the store gateway configurable

### LINKED ISSUE #289: [Feature request] Support for memcached as the index cache
**Is this a request for help?**:
No
---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
Feature request

I'd like to use Memcached instead of local memory for the store gateway. This is supported since [Thanos 0.10.0](https://github.com/thanos-io/thanos/releases/tag/v0.10.0). 

**Version of Helm and Kubernetes**:
Helm: `v2.16.1`
Kubernetes: `v1.12.8`

**Which chart in which version**:
prometheus-thanos `v2.10.0`

**What happened**:
The chart only supports the in-memory store

**What you expected to happen**:
Being able to configure the type of the index cache

**How to reproduce it** (as minimally and precisely as possible):

**Anything else we need to know**:
I'm going to raise a PR for this soon.

### PR DESCRIPTION
**Pull request by muffix**
#### What this PR does / why we need it:
Makes the index cache for the store gateway configurable. This now
supports the use of Memcached instead of local memory as the index cache
store (released in [Thanos 0.10.0](https://github.com/thanos-io/thanos/releases/tag/v0.10.0).

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #289

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://developercertificate.org) signed
- [x] Chart Version bumped (if the pr is an update to an existing chart)
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[fluentd-elasticsearch]`)

### CODE REVIEW

File: charts/prometheus-thanos/templates/store-gateway-statefulset.yaml [hunk 1]
```diff
@@ -44,7 +44,14 @@ spec:
           - store
           - --data-dir=/data
           - --log.level={{ .Values.storeGateway.logLevel }}
+          {{ if .Values.storeGateway.indexCacheConfig }}
+          - |
+            --index-cache.config=type: {{ .Values.storeGateway.indexCacheType }}
+            config:
+            {{- toYaml .Values.storeGateway.indexCacheConfig | nindent 14 }}
+          {{ else if .Values.storeGateway.indexCacheSize }}
```
Comment (axdotl) [inline charts/prometheus-thanos/templates/store-gateway-statefulset.yaml]: `indexCacheSize` was removed from `values.yaml`

File: charts/prometheus-thanos/templates/store-gateway-statefulset.yaml [hunk 2]
```diff
@@ -44,7 +44,14 @@ spec:
...
           - --index-cache-size={{ .Values.storeGateway.indexCacheSize }}
```
Comment (axdotl) [inline charts/prometheus-thanos/templates/store-gateway-statefulset.yaml]: Should we get rid of this config in favour of `index-cache.config`?

Comment (muffix) [reply charts/prometheus-thanos/templates/store-gateway-statefulset.yaml]: I thought I'd leae it in to support old values files, but happy to get rid of it if you thing that's better?

Comment (muffix) [reply charts/prometheus-thanos/templates/store-gateway-statefulset.yaml]: Updated. 👍

File: charts/prometheus-thanos/README.md [hunk 3]
```diff
@@ -233,9 +233,8 @@ The following table lists the configurable parameters of the prometheus-thanos c
 | `storeGateway.image.repository` | Docker image repo for store gateway | `quay.io/thanos/thanos` |
 | `storeGateway.image.tag` | Docker image tag for store gateway | `v0.10.1` |
 | `storeGateway.image.pullPolicy` | Docker image pull policy for store gateway | `IfNotPresent` |
-| `storeGateway.serviceAccount.create` | Create service account | `true` |
-| `storeGateway.serviceAccount.annotations` | Service account annotations | `nil` |
-| `storeGateway.indexCacheSize` | Index cache size | `500MB` |
+| `storeGateway.indexCacheConfig` | Type of the index cache, either `IN-MEMORY` or `MEMCACHED` | `IN-MEMORY` |
```
Comment (monotek) [inline charts/prometheus-thanos/README.md]: values & description for config & type seem twisted.

File: charts/prometheus-thanos/values.yaml [hunk 4]
```diff
@@ -106,7 +106,10 @@ storeGateway:
   # - name: GOOGLE_APPLICATION_CREDENTIALS
   #   value: /etc/gcp/secrets/credentials.json
   logLevel: info
-  indexCacheSize: 500MB
```
Comment (monotek) [inline charts/prometheus-thanos/values.yaml]: I'd like to see the config like this:

```
storeGateway:
  indexCache:
    config: 
    type: 
```
Also an example for memcached would be nice.

File: charts/prometheus-thanos/Chart.yaml [hunk 5]
```diff
@@ -2,7 +2,7 @@ apiVersion: v1
 appVersion: "0.10.1"
 description: A Helm chart for thanos monitoring components
 name: prometheus-thanos
-version: 2.10.0
+version: 2.11.0
```
Comment (monotek) [inline charts/prometheus-thanos/Chart.yaml]: If we remove cachesize variable we should rais the version to 3.0.0 as its a breaking change for users.

File: charts/prometheus-thanos/README.md (hunk 3 above)
Comment (muffix) [reply charts/prometheus-thanos/README.md]: Thanks, updated. 👍

File: charts/prometheus-thanos/values.yaml (hunk 4 above)
Comment (muffix) [reply charts/prometheus-thanos/values.yaml]: I think a full example would blow up the table a bit, so instead, I linked the Thanos docs in the comment which has a memcached example.

File: charts/prometheus-thanos/README.md [hunk 6]
```diff
@@ -233,9 +233,8 @@ The following table lists the configurable parameters of the prometheus-thanos c
 | `storeGateway.image.repository` | Docker image repo for store gateway | `quay.io/thanos/thanos` |
 | `storeGateway.image.tag` | Docker image tag for store gateway | `v0.10.1` |
 | `storeGateway.image.pullPolicy` | Docker image pull policy for store gateway | `IfNotPresent` |
-| `storeGateway.serviceAccount.create` | Create service account | `true` |
-| `storeGateway.serviceAccount.annotations` | Service account annotations | `nil` |
-| `storeGateway.indexCacheSize` | Index cache size | `500MB` |
+| `storeGateway.indexCache.config` | Config for the index cache, see [the docs](https://thanos.io/components/store.md/#index-cache) | `500MB` for the maximum size |
```
Comment (monotek) [inline charts/prometheus-thanos/README.md]: "`500MB` for the maximum size |" seems wrong there.

Comment (muffix) [reply charts/prometheus-thanos/README.md]: I think this replicates the behaviour of the previous verison of this chart which set `index-cache-size=500MB` (https://thanos.io/components/store.md/#in-memory-index-cache). Happy to change it to a more sensible value. Do you have anything in mind?
Or would you like me to be clearer on the description?

Comment (monotek) [reply charts/prometheus-thanos/README.md]: The description is good but teh default values "500MB for the maximum size" seems wrong.

Comment (muffix) [reply charts/prometheus-thanos/README.md]: Updated again. Thanks for having a look. 👍

**Pull request merged**